torch/_higher_order_ops/utils.py

```diff
@@ -651,3 +651,54 @@ def _tensor_storage(t) -> StorageWeakRef:
             if isinstance(inp, torch.Tensor) and _tensor_storage(inp) in out_storage_map
         }
         return mutated_inputs, inp_inp_alias_map, inp_out_alias_map, out_out_alias_map
+
+
+registered_hop_fake_fns: dict[torch._ops.OpOverload, Callable] = {}
+
+
+def register_hop_fake(hop, fn=None):
+    """
+    Register a fake function for a HOP. This is conceptually equivalent of the
+    register_fake utility for the custom ops. The registered function is called
+    inside the fake_tensor _dispatch_impl.
+    """
+    assert hop not in registered_hop_fake_fns
+
+    def register(func):
+        from torch._subclasses.fake_tensor import FakeTensorMode
+
+        # Redirect the hop to the fake tensor mode implementation.
+        @hop.py_impl(FakeTensorMode)
+        def _(mode, *args, **kwargs):
+            return mode.__torch_dispatch__(hop, [], args, kwargs)
+
+        registered_hop_fake_fns[hop] = func
+        return func
+
+    if fn is None:
+        return register
+    return register(fn)
+
+
+class FunctionalizeCtxWrapper:
+    """
+    This is a dummy wrapper to facilitate fake tensor caching.
+
+    For AOT Dispatcher metadata collection pass, HOPs go from functionalization
+    key to fake tensor key. The functionalization key wraps the subgraphs in a
+    function, which changes from call to call even though the subgraph might
+    still be same.
+
+    To enable fake tensor caching, we just wrap the ctx and subgraph in this
+    class and then use the subgraph as the hash.
+    """
+
+    def __init__(self, ctx, subgraph):
+        self.ctx = ctx
+        self.subgraph = subgraph
+
+    def __hash__(self):
+        return id(self.subgraph)
+
```

Comment:
nit: Can you add some sort of repr so we know what we're dealing with when we're debugging?

Reply:
Added.